app/actions/AuthorisedAction.scala

```diff
@@ -129,75 +129,84 @@ class AuthorisedAction @Inject()(appConfig: AppConfig,
-        Future.successful(Redirect(controllers.errors.routes.YouNeedAgentServicesController.show))
-    }
-  }
+      private def handleForValidAgent[A](block: AuthorisationRequest[A] => Future[Result],
+                                         mtdItId: String,
+                                         nino: String,
+                                         enrolments: Enrolments,
+                                         isSupportingAgent: Boolean)
+                                        (implicit request: Request[A], hc: HeaderCarrier): Future[Result] = {
+        isSupportingAgent match {
+          case true =>
+            logger.warn(s"$agentAuthLogString - Secondary agent unauthorised")
+            Future.successful(Redirect(controllers.errors.routes.SupportingAgentAuthErrorController.show))
+          case false =>
+            enrolmentGetIdentifierValue(EnrolmentKeys.Agent, EnrolmentIdentifiers.agentReference, enrolments) match {
+              case Some(arn) =>
+                sessionIdBlock(
+                  errorLogString = s"$agentAuthLogString - No session id in request",
+                  errorAction = Future.successful(Redirect(appConfig.signInUrl))
+                )(sessionId =>
+                  block(AuthorisationRequest(
+                    user = User(mtdItId, Some(arn), nino, AffinityGroup.Agent.toString, sessionId, isSupportingAgent),
+                    request = request
+                  ))
+                )
+              case None =>
+                logger.warn(s"$agentAuthLogString - Agent with no HMRC-AS-AGENT enrolment. Rendering unauthorised view.")
+                Future.successful(Redirect(controllers.errors.routes.YouNeedAgentServicesController.show))
+            }
+        }
+      }
 
-  private[actions] def agentAuthentication[A](block: AuthorisationRequest[A] => Future[Result])
-                                             (implicit request: Request[A], hc: HeaderCarrier): Future[Result] = {
-    val optionalNino = request.session.get(SessionValues.CLIENT_NINO)
-    val optionalMtdItId = request.session.get(SessionValues.CLIENT_MTDITID)
-
-    (optionalMtdItId, optionalNino) match {
-      case (Some(mtdItId), Some(nino)) =>
-        authService
-          .authorised(agentAuthPredicate(mtdItId))
-          .retrieve(allEnrolments)(
-            enrollments => handleForValidAgent(block, mtdItId, nino, enrollments, isSupportingAgent = false)
-          )
-          .recoverWith(agentRecovery(block, mtdItId, nino))
-      case (mtdItId, nino) =>
-        logger.info(s"[AuthorisedAction][agentAuthentication] - Agent does not have session key values. " +
-          s"Redirecting to view & change. MTDITID missing:${mtdItId.isEmpty}, NINO missing:${nino.isEmpty}")
-        Future.successful(Redirect(appConfig.viewAndChangeEnterUtrUrl))
-    }
+
+
+      private[actions] def agentAuthentication[A](block: AuthorisationRequest[A] => Future[Result])
```

Comment:
A lot of indents in this file below agentRecovery have been broken too, could they be aligned again e.g. handleForValidAgent